lambdas/package/index.py

```diff
@@ -0,0 +1,210 @@
+"""
+Provide a virtual-file-system view of a package's logical keys.
+"""
+
+import io
+import json
+
+import boto3
+import pandas as pd
+
+from t4_lambda_shared.decorator import api, validate
+from t4_lambda_shared.utils import get_default_origins, make_json_response
+
+S3_DOMAIN_SUFFIX = '.amazonaws.com'
+
+SCHEMA = {
+    'type': 'object',
+    'properties': {
+        'bucket': {
+            'type': 'string'
+        },
+        'manifest': {
+            'type': 'string'
+        },
+        'access_key': {
+            'type': 'string'
+        },
+        'secret_key': {
+            'type': 'string'
+        },
+        'prefix': {
+            'type': 'string'
+        },
+        'logical_key': {
+            'type': 'string'
+        }
+    },
+    'required': ['bucket', 'manifest', 'access_key', 'secret_key'],
+    'additionalProperties': False
+}
+
+
+class IncompleteResultException(Exception):
+    """
+    Exception indicating an incomplete response
+    (e.g., from S3 Select)
+    """
+
+
+def load_df(s3response):
+    """
+    Read a streaming response from s3 select into a
+    Pandas DataFrame
+    """
+    buffer = io.StringIO()
+    end_event_received = False
+    stats = None
+    for event in s3response['Payload']:
+        if 'Records' in event:
+            records = event['Records']['Payload'].decode()
+            buffer.write(records)
+        elif 'Progress' in event:
+            print(event['Progress']['Details'])
+        elif 'Stats' in event:
+            stats = event['Stats']['Details']
+        elif 'End' in event:
+            # End event indicates that the request finished successfully
+            end_event_received = True
+
+    if not end_event_received:
+        raise IncompleteResultException("Error: Received an incomplete response from S3 Select.")
+
+    buffer.seek(0)
+    df = pd.read_json(buffer, lines=True)  # pylint: disable=invalid-name
+    return df, stats
+
+
+def load_manifest_detail(s3response):
+    """
+    Read a streaming response from s3 select into a dict
+    """
+    end_event_received = False
+    response_data = None
+    for event in s3response['Payload']:
+        if 'Records' in event:
+            records = event['Records']['Payload'].decode()
+            response_data = json.loads(records)
```

Comment:
Probably this is unlikely, but [it can contain only partial record](https://docs.aws.amazon.com/AmazonS3/latest/API/RESTSelectObjectAppendix.html#RESTSelectObjectAppendix-responses):

> Records message: Can contain a single record, partial records, or multiple records.

So I think we have to iterate till the end here too.

Reply:
You mean that we would see a payload with only a subset of a single row?

Reply:
That's not clear for me from reading the mentioned doc if it's possible for this specific case, but I think doesn't make much sense to not handle it in a common way.